core/utils/wei.go

```diff
@@ -198,3 +205,78 @@ func (w *Wei) UnmarshalText(b []byte) error {
 	}
 	return errors.Errorf("unable to parse %q", s)
 }
+
+func (w *Wei) ToInt() *big.Int {
+	return (*big.Int)(w)
+}
+
+func (w *Wei) Int64() int64 {
+	return w.ToInt().Int64()
+}
+
+func (w *Wei) Cmp(y *Wei) int {
+	return w.ToInt().Cmp(y.ToInt())
+}
+
+func (w *Wei) Equal(y *Wei) bool {
+	return w.Cmp(y) == 0
+}
+
+type ErrNegative struct{ InvalidVal *big.Int }
+
+func (e ErrNegative) Error() string {
+	return fmt.Sprintf("invalid Wei value: %s", e.InvalidVal.String())
+}
+
+func WeiMax(x, y *Wei) *Wei {
+	return NewWei(bigmath.Max(x.ToInt(), y.ToInt()))
+}
+
+func WeiMin(x, y *Wei) *Wei {
+	return NewWei(bigmath.Min(x.ToInt(), y.ToInt()))
+}
+
+// NOTE: Maths functions always return newly allocated number and do not mutate
+
+func (w *Wei) Sub(y *Wei) (*Wei, error) {
+	result := big.NewInt(0).Sub(w.ToInt(), y.ToInt())
+	if result.Cmp(big.NewInt(0)) < 0 {
+		return nil, ErrNegative{InvalidVal: result}
```

Comment:
Why is this an error? Shouldn't it be left up to the caller whether or not a negative value is invalid?

Reply:
Yeah not sure why I did this probably was trying to be too clever. Will remove.

Reply:
I could see a case for non-negative validation during TOML unmarshalling, with an 'unsigned wei' type or something :shrug: :thinking: